integration-cli: readContainerFileWithExec for links tests

ahmetb opened the pull request:
Shout out to @estesp for the idea. Some use cases of
`readContainerFile` can be replaced with `docker exec $id cat $file`.
This helper method can eliminate the assumption that
host/cli should be on the same machine, partially.

TestRunMutableNetworkFiles and TestRunResolvconfUpdater still
need to access the docker host filesystem as they modify
the file directly from there assuming cli and daemon are
on the same machine. They may keep using`readContainerFile`.

This fixes `TestLinksUpdateOnRestart` and `TestLinksHostsFilesInject`
for Windows CI.

Signed-off-by: Ahmet Alp Balkan ahmetalpbalkan@gmail.com
Label: `#windows`
Cc: @estesp @duglin @unclejack @jfrazelle @tianon

ahmetb committed:
readContainerFileWithExec for links tests

Shout out to @estesp for the idea. Some use cases of
`readContainerFile` can be replaced with `docker exec $id cat $file`.
This helper method can eliminate the requirement that
host/cli should be on the same machine.

TestRunMutableNetworkFiles and TestRunResolvconfUpdater still
need to access the docker host filesystem as they modify
the file directly from there assuming cli and daemon are
on the same machine.

This fixes TestLinksUpdateOnRestart and TestLinksHostsFilesInject
for Windows CI.

Signed-off-by: Ahmet Alp Balkan <ahmetalpbalkan@gmail.com>

ahmetalpbalkan commented:
> https://jenkins.dockerproject.com/job/Windows-PRs/46/console
=== RUN TestLinksUpdateOnRestart
[PASSED]: link - ensure containers hosts files are updated on restart
=== RUN TestLinksHostsFilesInject
[PASSED]: link - ensure containers hosts files are updated with the link alias. :+1:

jfrazelle commented:
this is going to break lxc, can we move all the exec tests to a certain file that wont run on lxc, see the exec_test file for example

jfrazelle commented:
or we need a way to do this without exec

ahmetalpbalkan commented:
Or we may skip these altogether or move to "//+build daemon" as it's not really testing the CLI but daemon.

cpuguy83 commented:
Or use `POST /contaienrs/<id>/copy`

ahmetalpbalkan commented:
@cpuguy83 erm, you're right... @jfrazelle how's that sound to you?

jfrazelle commented:
sgtm

LK4D4 commented:
yeah, seems like copy is best option (apart from dropping lxc support)

ahmetalpbalkan commented:
@jfrazelle @LK4D4 I pretty much implemented the `docker cp` support except it appears that we can't cp "/etc/hosts" from a container #9998, #10037. @tianon 

How about we skip these 2 test cases with `testRequires(t, ExecSupport)` and keep the original "`docker exec .. cat /file`" implementation? 

(I guess we can use `test_no_exec` go build tag to find out if execdriver==lxc and create 2 files test_vars_exec.go, test_vars_noexec.go and add `supportsExec` const for `testRequires`? @jfrazelle)

LK4D4 commented:
+1 for `supportExec`

ahmetb committed:
integration-cli: add test requirement ExecSupport

Skip tests based on remote daemon's exec support (to exclude
these tests from `make test` ran in LXC case). Makes use of
`test_no_exec` build tag passed by build scripts.

Signed-off-by: Ahmet Alp Balkan <ahmetalpbalkan@gmail.com>

ahmetalpbalkan commented:
@LK4D4 @cpuguy83 @jfrazelle I just took the fixes. 
- Made use of `test_no_exec` build tag to skip the tests in LXC case (102e061)
- implemented `readContainerFileWithExec` (6bbb456).

jfrazelle commented:
yayyy thanks @ahmetalpbalkan LGTM

jfrazelle commented:
ping @LK4D4

crosbymichael commented:
LGTM

ahmetalpbalkan commented:
yaay, thanks. :joy:
